text/0000-ques-in-main.md

```diff
@@ -0,0 +1,727 @@
+environment where there is no such thing as an "error message",
+`io::stderr()` probably already points to the bit bucket, so `report`
+functions can go ahead and use `eprintln!` anyway.
+
+There are also environments where
+[returning from `main` constitutes a _bug_.][divergent-main] If you
+are implementing an operating system kernel, for instance, there may
+be nothing to return to.  Then you want it to be a compile-time error
+for `main` to return anything other than `!`.  If everything is
+implemented correctly, such environments should be able to get that
+effect by omitting all stock impls of `Termination` other than for
+`!`.  Perhaps there should also be a compiler hook that allows such
+environments to refuse to let you impl Termination yourself.
+
+[divergent-main]: https://internals.rust-lang.org/t/allowing-for-main-to-be-divergent-in-embedded-environments/4717
+
+## Test functions and doctests
+[test-functions-and-doctests]: #test-functions-and-doctests
+
+The harness for `#[test]` functions will need to be changed, similarly
+to how the "start" lang item was changed.  Tests which return
+anything whose `report()` method returns a nonzero value should be
+considered to have failed.
+
+Doctests require a little magic in rustdoc: when `maketest` sees that
+it needs to insert a function head for `main`, it should now write out
+
+``` rust
+fn main () -> Result<(), ErrorT> {
+   ...
+   Ok(())
+}
+```
+
+for some value of `ErrorT` TBD.  It doesn't need to parse the body of
+the test to know whether it should do this; it can just do it
+unconditionally.
+
+## New constructors for `ExitStatus`
+[exit-status]: #exit-status
+
+As mentioned above, we propose to reuse `process::ExitStatus` as a way
+for a program to generate both success and failure conditions
+_without_ any errors printed, by returning it from `main`.  To make
+this more convenient, we also propose to add the following new
+constructors to `ExitStatus`:
+
+``` rust
+impl ExitStatus {
+    /// Return an ExitStatus value representing success.
+    /// (ExitStatus::ok()).success() is guaranteed to be true, and
+    /// (ExitStatus::ok()).code() is guaranteed to be Some(0).
+    pub fn ok() -> Self;
+
+    /// Return an ExitStatus value representing failure.
+    /// (ExitStatus::failure()).success() is guaranteed to be false, and
+    /// (ExitStatus::failure()).code() is guaranteed to be Some(n),
+    /// for some unspecified n, 1 < n < 64.
+    pub fn failure() -> Self;
+
+    /// Return an ExitStatus value representing a specific exit code.
+    /// The difference between this method and ExitStatusExt::from_raw
+    /// is that this method can only be used to produce ExitStatus
+    /// values that will pass unmodified through the operating system
+    /// primitive "exit" and "wait" operations on all supported
+    /// platforms.  (Conveniently, this is exactly the range of u8.)
+    pub fn from_code(code: u8) -> Self;
```

Comment:
the `u8` here is a bad smell, an unnecessary limitation. I do read your explanations here and below. but ... os dependent implementation detail should not affect interface.